File: apps/windows/TeslaSync.App/feature-views/VehicleSettingsTab/VehicleSettingsTab.ViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews.Vehicles;

/// <summary>The tone of a transient notice raised for the toast surface (web <c>useMutationToast</c>).</summary>
public enum VehicleSettingsTabNoticeKind
{
    /// <summary>A successful mutation (web <c>toast.success</c>).</summary>
    Success,

    /// <summary>A failed mutation (web <c>toast.error</c>).</summary>
    Error,
}

/// <summary>A localized transient notice for the toast surface (web mutation <c>onSuccess</c> / <c>onError</c>).</summary>
public sealed record VehicleSettingsTabNotice(VehicleSettingsTabNoticeKind Kind, string Message);

/// <summary>
/// UI-thread-free state holder for one editable per-vehicle setting row — the native port of the web
/// <c>VehicleSettingRow</c> sub-component. It seeds an editable <see cref="Draft"/> from the resolved effective value
/// (re-seeding only when the effective value actually changes — the web <c>useEffect([initialDraft])</c> guard, which
/// keeps a field's edits while a background refresh runs), exposes the dirty diff that arms the Save button and the
/// override flag that arms the Reset button, and carries the projected <see cref="Display"/> (label, help, source
/// pill). All mutation orchestration lives on the parent <see cref="VehicleSettingsTabViewModel"/>; this row only
/// holds the per-row edit state.
/// </summary>
public sealed class VehicleSettingRowViewModel : INotifyPropertyChanged
{
    private readonly ILocalizer _localizer;

    private string _initialDraft = string.Empty;
    private string _draft = string.Empty;
    private string? _validationError;
    private bool _isSaving;
    private bool _isResetting;
    private VehicleSettingRowDisplay _display;

    /// <summary>Creates the row over its descriptor and the i18n facade, seeded from an (optional) effective value.</summary>
    /// <param name="descriptor">The static per-key descriptor (kind, options, length).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="effective">The resolved effective value, or null when the key has no resolved row.</param>
    public VehicleSettingRowViewModel(VehicleSettingDescriptor descriptor, ILocalizer localizer, EffectiveSettingData? effective)
    {
        ArgumentNullException.ThrowIfNull(descriptor);
        ArgumentNullException.ThrowIfNull(localizer);

        Descriptor = descriptor;
        _localizer = localizer;
        _display = VehicleSettingsTabProjection.ProjectRow(descriptor, effective, localizer);
        _initialDraft = _display.InitialDraft;
        _draft = _initialDraft;
    }

    /// <inheritdoc />
    public event PropertyChangedEventHandler? PropertyChanged;

    /// <summary>The static per-key descriptor (kind, options, max length, autocomplete).</summary>
    public VehicleSettingDescriptor Descriptor { get; }

    /// <summary>The setting key (web <c>descriptor.key</c>).</summary>
    public string Key => Descriptor.Key;

    /// <summary>The projected, render-ready row chrome (label, help, source pill, automation ids).</summary>
    public VehicleSettingRowDisplay Display
    {
        get => _display;
        private set => Set(ref _display, value);
    }

    /// <summary>The current editable draft (always a string so every kind renders uniformly).</summary>
    public string Draft
    {
        get => _draft;
        set
        {
            if (Set(ref _draft, value ?? string.Empty))
            {
                Raise(nameof(IsDirty));
                Raise(nameof(CanSave));
            }
        }
    }

    /// <summary>The inline validation error to show under the input, or null when valid (web <c>validationError</c>).</summary>
    public string? ValidationError
    {
        get => _validationError;
        private set
        {
            if (Set(ref _validationError, value))
            {
                Raise(nameof(HasValidationError));
            }
        }
    }

    /// <summary>True while this row's upsert is in flight (web <c>upsert.isPending</c>).</summary>
    public bool IsSaving
    {
        get => _isSaving;
        internal set
        {
            if (Set(ref _isSaving, value))
            {
                Raise(nameof(CanSave));
            }
        }
    }

    /// <summary>True while this row's reset is in flight (web <c>reset.isPending</c>).</summary>
    public bool IsResetting
    {
        get => _isResetting;
        internal set
        {
            if (Set(ref _isResetting, value))
            {
                Raise(nameof(CanReset));
            }
        }
    }

    /// <summary>True when the draft differs from the seeded effective value (web <c>dirty</c>).</summary>
    public bool IsDirty => !string.Equals(_draft, _initialDraft, StringComparison.Ordinal);

    /// <summary>True when an inline validation error is present.</summary>
    public bool HasValidationError => _validationError is not null;

    /// <summary>True when the resolved value is a per-vehicle override (the only source that can be reset).</summary>
    public bool IsOverride => _display.IsOverride;

    /// <summary>True when the Save button is enabled (dirty and not already saving).</summary>
    public bool CanSave => IsDirty && !_isSaving;

    /// <summary>True when the Reset button is enabled (an override that is not already resetting).</summary>
    public bool CanReset => IsOverride && !_isResetting;

    /// <summary>The seeded draft the dirty diff compares against (the web <c>initialDraft</c>).</summary>
    public string InitialDraft => _initialDraft;

    /// <summary>
    /// Fold a new resolved effective value into the row. The source pill / display always refreshes; the editable
    /// draft is re-seeded (and the validation error cleared) only when the seeded value actually changes — the web
    /// <c>useEffect([initialDraft])</c> dependency, so a field keeps its in-progress edits during a background refresh.
    /// </summary>
    public void UpdateEffective(EffectiveSettingData? effective)
    {
        Display = VehicleSettingsTabProjection.ProjectRow(Descriptor, effective, _localizer);
        Raise(nameof(IsOverride));
        Raise(nameof(CanReset));

        string nextInitial = Display.InitialDraft;
        if (!string.Equals(nextInitial, _initialDraft, StringComparison.Ordinal))
        {
            _initialDraft = nextInitial;
            ValidationError = null;
            Draft = nextInitial;
            Raise(nameof(InitialDraft));
        }
        else
        {
            Raise(nameof(IsDirty));
            Raise(nameof(CanSave));
        }
    }

    /// <summary>Clear the inline validation error at the start of a save attempt (web <c>setValidationError(null)</c>).</summary>
    internal void ClearValidationError() => ValidationError = null;

    /// <summary>Surface an inline validation error (web <c>setValidationError(...)</c>).</summary>
    internal void SetValidationError(string message) => ValidationError = message;

    private bool Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
        {
            return false;
        }

        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        return true;
    }

    private void Raise(string name) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

/// <summary>
/// UI-thread-free state holder backing the WinUI <c>VehicleSettingsTab</c> view — the native port of the web
/// component (web/src/features/vehicles/components/VehicleSettingsTab.tsx). It streams the cache-then-network
/// per-vehicle settings (the web <c>useVehicleSettings</c> read) through the injected
/// <see cref="IVehicleSettingsTabSource"/>, folds every emission into the mutually-exclusive <see cref="State"/>
/// (loading / loaded / empty / error / stale / offline) and the header freshness flags, and keeps one stable
/// <see cref="VehicleSettingRowViewModel"/> per descriptor whose effective value is refreshed in place. The per-key
/// <see cref="SaveRowAsync"/> / <see cref="ResetRowAsync"/> reproduce the web upsert / reset mutations — validating
/// the draft, surfacing the toast and re-reading the resolver so the saved value flows back (the web
/// <c>invalidateAndBroadcast</c>). Drive it from one confinement (the UI thread); it is not internally synchronised.
/// </summary>
public sealed class VehicleSettingsTabViewModel : INotifyPropertyChanged, IDisposable
{
    private readonly IVehicleSettingsTabSource _source;
    private readonly ILocalizer _localizer;
    private readonly VehicleSettingsTabDiagnostics _diagnostics;
    private readonly Func<DateTimeOffset> _clock;
    private readonly long _vehicleId;
    private readonly IReadOnlyList<VehicleSettingRowViewModel> _rows;

    private CancellationTokenSource? _cts;
    private bool _disposed;

    private VehicleSettingsData _settings = VehicleSettingsData.Empty;
    private VehicleSettingsTabState _state = VehicleSettingsTabState.Loading;
    private VehicleSettingsTabDisplay _display;
    private DateTimeOffset? _updatedAt;
    private bool _isFetching;
    private bool _isError;
    private bool _isStale;
    private string? _errorMessage;
    private int _attempts;

    /// <summary>Creates the holder over its data source, localizer, vehicle id and (optional) clock / diagnostics.</summary>
    /// <param name="source">The per-vehicle settings data port (read + upsert + reset).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="vehicleId">The vehicle id from the route (web <c>vehicleId</c> prop).</param>
    /// <param name="clock">Injectable clock for deterministic freshness in tests.</param>
    /// <param name="diagnostics">The PII-safe diagnostics sink for the <c>view.opened</c> event.</param>
    public VehicleSettingsTabViewModel(
        IVehicleSettingsTabSource source,
        ILocalizer localizer,
        long vehicleId,
        Func<DateTimeOffset>? clock = null,
        VehicleSettingsTabDiagnostics? diagnostics = null)
    {
        ArgumentNullException.ThrowIfNull(source);
        ArgumentNullException.ThrowIfNull(localizer);

        _source = source;
        _localizer = localizer;
        _vehicleId = vehicleId;
        _clock = clock ?? (() => DateTimeOffset.Now);
        _diagnostics = diagnostics ?? new VehicleSettingsTabDiagnostics();

        _rows = VehicleSettingDescriptor.All
            .Select(descriptor => new VehicleSettingRowViewModel(descriptor, localizer, _settings.Find(descriptor.Key)))
            .ToArray();
        _display = VehicleSettingsTabProjection.Project(_state, _settings, localizer);
    }

    /// <inheritdoc />
    public event PropertyChangedEventHandler? PropertyChanged;

    /// <summary>Raised with a localized transient notice for the toast surface (web <c>useMutationToast</c>).</summary>
    public event EventHandler<VehicleSettingsTabNotice>? NoticeRequested;

    /// <summary>The current top-level data state (loading / loaded / empty / error / stale / offline).</summary>
    public VehicleSettingsTabState State
    {
        get => _state;
        private set => Set(ref _state, value);
    }

    /// <summary>The projected, render-ready surface chrome the view binds to.</summary>
    public VehicleSettingsTabDisplay Display
    {
        get => _display;
        private set => Set(ref _display, value);
    }

    /// <summary>The stable per-descriptor row state holders (one per whitelist key, in render order).</summary>
    public IReadOnlyList<VehicleSettingRowViewModel> Rows => _rows;

    /// <summary>True while a (re)fetch is in flight (the header freshness chip pulses).</summary>
    public bool IsFetching
    {
        get => _isFetching;
        private set => Set(ref _isFetching, value);
    }

    /// <summary>True when the last load failed with no cache (drives the error surface + header chip).</summary>
    public bool IsError
    {
        get => _isError;
        private set => Set(ref _isError, value);
    }

    /// <summary>True when the shown payload is older than the freshness window (stale or offline).</summary>
    public bool IsStale
    {
        get => _isStale;
        private set => Set(ref _isStale, value);
    }

    /// <summary>The localized error / offline message (web <c>ErrorDisplay</c> text), or null when healthy.</summary>
    public string? ErrorMessage
    {
        get => _errorMessage;
        private set => Set(ref _errorMessage, value);
    }

    /// <summary>Last successful update timestamp surfaced in the header freshness chip.</summary>
    public DateTimeOffset? UpdatedAt
    {
        get => _updatedAt;
        private set => Set(ref _updatedAt, value);
    }

    /// <summary>Number of load attempts started (including retries).</summary>
    public int Attempts
    {
        get => _attempts;
        private set => Set(ref _attempts, value);
    }

    /// <summary>The vehicle id this holder is bound to.</summary>
    public long VehicleId => _vehicleId;

    /// <summary>Record that the surface was opened (PII-safe diagnostics).</summary>
    public void NotifyOpened() => _diagnostics.RecordViewOpened();

    /// <summary>
    /// Run (or re-run) the cache-then-network settings load: counts the attempt, shows the skeleton only when nothing
    /// is already visible (otherwise keeps the rows while refreshing), and folds every emission into
    /// <see cref="State"/>. A superseding load cancels the prior one.
    /// </summary>
    public async Task LoadAsync(CancellationToken cancellationToken = default)
    {
        var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        var previous = Interlocked.Exchange(ref _cts, cts);
        previous?.Cancel();
        previous?.Dispose();

        Attempts++;
        if (!HasContent())
        {
            SetLoading();
        }
        else
        {
            IsFetching = true;
            Reproject();
        }

        try
        {
            await foreach (var result in _source.StreamSettingsAsync(_vehicleId, cts.Token).ConfigureAwait(false))
            {
                Apply(result);
            }
        }
        catch (OperationCanceledException)
        {
            // Superseded by a newer load (or disposed) — drop this emission silently.
        }
    }

    /// <summary>Refresh the settings (web query refetch).</summary>
    public Task RefreshAsync(CancellationToken cancellationToken = default) => LoadAsync(cancellationToken);

    /// <summary>Retry after a failure — re-runs the load from the top (web <c>ErrorDisplay onRetry</c>).</summary>
    public Task RetryAsync(CancellationToken cancellationToken = default) => LoadAsync(cancellationToken);

    /// <summary>
    /// Save one row (web <c>handleSave</c> + <c>useUpsertVehicleSetting</c>): clears any prior error, validates the
    /// draft (empty → "required"; invalid → the kind-specific message), then upserts the typed value. On success it
    /// raises the saved toast and re-reads the resolver so the committed value flows back; on failure it raises the
    /// error toast and keeps the edited draft for a retry. Concurrent saves of the same row are ignored.
    /// </summary>
    public async Task SaveRowAsync(VehicleSettingRowViewModel row, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(row);
        if (_disposed || row.IsSaving)
        {
            return;
        }

        row.ClearValidationError();

        var parsed = VehicleSettingDraft.ParseDraft(row.Descriptor, row.Draft);
        switch (parsed.Status)
        {
            case VehicleSettingParseStatus.Empty:
                row.SetValidationError(_display.RequiredMessage);
                return;
            case VehicleSettingParseStatus.Invalid:
                row.SetValidationError(_localizer.GetString(parsed.MessageKey!, parsed.Fallback!));
                return;
        }

        row.IsSaving = true;
        try
        {
            await _source.UpsertAsync(_vehicleId, row.Key, parsed.Value!, cancellationToken).ConfigureAwait(false);
            RaiseNotice(VehicleSettingsTabNoticeKind.Success, _display.SavedToast);
            await LoadAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            // Superseded / disposed — leave the draft intact for a later retry.
        }
        catch (Exception)
        {
            RaiseNotice(VehicleSettingsTabNoticeKind.Error, _display.SaveFailedToast);
        }
        finally
        {
            row.IsSaving = false;
        }
    }

    /// <summary>
    /// Reset one row to its default (web <c>handleReset</c> + <c>useResetVehicleSetting</c>): a no-op unless the row
    /// is a per-vehicle override. On success it raises the reverted toast and re-reads the resolver; on failure it
    /// raises the error toast. Concurrent resets of the same row are ignored.
    /// </summary>
    public async Task ResetRowAsync(VehicleSettingRowViewModel row, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(row);
        if (_disposed || row.IsResetting || !row.IsOverride)
        {
            return;
        }

        row.IsResetting = true;
        try
        {
            await _source.ResetAsync(_vehicleId, row.Key, cancellationToken).ConfigureAwait(false);
            RaiseNotice(VehicleSettingsTabNoticeKind.Success, _display.ResetToast);
            await LoadAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            // Superseded / disposed.
        }
        catch (Exception)
        {
            RaiseNotice(VehicleSettingsTabNoticeKind.Error, _display.ResetFailedToast);
        }
        finally
        {
            row.IsResetting = false;
        }
    }

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        var cts = Interlocked.Exchange(ref _cts, null);
        cts?.Cancel();
        cts?.Dispose();
        GC.SuppressFinalize(this);
    }

    private bool HasContent() =>
        _state is VehicleSettingsTabState.Loaded
            or VehicleSettingsTabState.Empty
            or VehicleSettingsTabState.Stale
            or VehicleSettingsTabState.Offline;

    private void Apply(RepositoryResult<VehicleSettingsData> result)
    {
        switch (result.Status)
        {
            case LoadStatus.Loading:
                if (!HasContent())
                {
                    SetLoading();
                }

                IsFetching = true;
                Reproject();
                break;

            case LoadStatus.Cached:
                ApplySnapshot(result.Value!, result.FetchedAt, result.IsStale, fetching: false, offline: false);
                break;

            case LoadStatus.Refreshing:
                ApplySnapshot(result.Value!, result.FetchedAt, result.IsStale, fetching: true, offline: false);
                break;

            case LoadStatus.Loaded:
                ApplySnapshot(result.Value!, result.FetchedAt, stale: false, fetching: false, offline: false);
                break;

            case LoadStatus.Empty:
                SetEmpty(result.FetchedAt);
                break;

            case LoadStatus.Offline:
                ApplySnapshot(result.Value!, result.FetchedAt, stale: true, fetching: false, offline: true);
                break;

            default:
                SetError();
                break;
        }
    }

    private void ApplySnapshot(
        VehicleSettingsData settings,
        DateTimeOffset? fetchedAt,
        bool stale,
        bool fetching,
        bool offline)
    {
        _settings = settings;
        UpdateRows(settings);
        UpdatedAt = fetchedAt;
        IsFetching = fetching;
        IsStale = stale;
        IsError = false;
        ErrorMessage = offline ? _localizer.GetString("vehicleSettings.error", "Could not load vehicle settings.") : null;

        State = offline
            ? VehicleSettingsTabState.Offline
            : stale
                ? VehicleSettingsTabState.Stale
                : VehicleSettingsTabState.Loaded;

        Reproject();
    }

    private void SetLoading()
    {
        IsError = false;
        ErrorMessage = null;
        State = VehicleSettingsTabState.Loading;
        Reproject();
    }

    private void SetEmpty(DateTimeOffset? fetchedAt)
    {
        // The resolver always returns the full key whitelist, so an empty payload is unusual; the rows still render
        // with their system defaults (web maps the static descriptor list regardless of the payload).
        _settings = VehicleSettingsData.Empty;
        UpdateRows(_settings);
        UpdatedAt = fetchedAt;
        IsFetching = false;
        IsStale = false;
        IsError = false;
        ErrorMessage = null;
        State = VehicleSettingsTabState.Empty;
        Reproject();
    }

    private void SetError()
    {
        IsFetching = false;
        IsStale = false;
        IsError = true;
        // Web passes a fixed `t('vehicleSettings.error')` to ErrorDisplay regardless of the underlying failure.
        ErrorMessage = _localizer.GetString("vehicleSettings.error", "Could not load vehicle settings.");
        State = VehicleSettingsTabState.Error;
        Reproject();
    }

    private void UpdateRows(VehicleSettingsData settings)
    {
        foreach (var row in _rows)
        {
            row.UpdateEffective(settings.Find(row.Key));
        }
    }

    private void Reproject() => Display = VehicleSettingsTabProjection.Project(_state, _settings, _localizer);

    private void RaiseNotice(VehicleSettingsTabNoticeKind kind, string message) =>
        NoticeRequested?.Invoke(this, new VehicleSettingsTabNotice(kind, message));

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
        {
            return;
        }

        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/VehicleSettingsTab/VehicleSettingsTabViewModel.kt
// UI-thread-free state holder backing the Compose [VehicleSettingsTab] surface — the native port of the hook
// composition the web component owns (web/src/features/vehicles/components/VehicleSettingsTab.tsx +
// web/src/api/hooks/useVehicleSettings.ts). It binds the shared P1/S8 layer through
// [VehicleSettingsTabSource]: it collects the cache-then-network resolver feed, holds the user's per-key
// in-progress draft edits + inline validation (the web `VehicleSettingRow` local `useState`), and runs the
// per-key override save (web `useUpsertVehicleSetting`) and reset-to-default (web `useResetVehicleSetting`).
// The view never performs HTTP — it only collects [state] and calls the trigger methods.
//
// The single combined [state] folds five inputs — the settings feed, the per-key draft map, the per-key
// validation map, and the per-key saving / resetting in-flight sets — which the pure
// [VehicleSettingsTabProjection] (in the composable) turns into the render-ready display, covering every
// state the surface draws: loading, the editable rows, a hard error + retry, and the stale/offline envelope.
// A successful mutation clears that key's draft + validation and re-fetches the feed (so the row follows the
// new effective value and stops being dirty), then emits a one-shot success toast; a failure refreshes
// nothing and emits an error toast — the web `onSuccess`/`onError` split. Toasts are surfaced as one-shot
// [io.teslasync.android.data.UiEvent]s (the render boundary owns the string lookup), never held as state.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/feature-views/VehicleSettingsTab) cannot form a valid Kotlin package.
@file:Suppress("InvalidPackageDeclaration")

package io.teslasync.android.featureviews.vehiclesettingstab

import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import io.teslasync.android.data.BaseFeedViewModel
import io.teslasync.android.data.UiEvent
import io.teslasync.android.data.errorKindOf
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.vehiclesettings.EffectiveSettingSource
import io.teslasync.shared.core.presentation.vehiclesettings.findEffectiveSetting
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update

/**
 * State holder backing the Compose [VehicleSettingsTab].
 *
 * @param source the cache-then-network per-vehicle settings seam (a shared S8 adapter in production, a fake
 *   in tests). The view-model owns no networking — it only collects + re-shares the feed and runs mutations.
 * @param logger the single sanctioned redacting logger (ADR-016); receives the PII-safe `view.opened`,
 *   `vehicleSettings.save`, `vehicleSettings.reset`, and `vehicleSettings.refresh` events — never a value.
 * @param scope test seam; production passes nothing and uses `viewModelScope`.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class VehicleSettingsTabViewModel(
    private val source: VehicleSettingsTabSource,
    logger: Logger,
    scope: CoroutineScope? = null,
) : BaseFeedViewModel(logger, scope) {
    private val drafts = MutableStateFlow<Map<String, String>>(emptyMap())
    private val validation = MutableStateFlow<Map<String, VehicleSettingValidation>>(emptyMap())
    private val savingKeys = MutableStateFlow<Set<String>>(emptySet())
    private val resettingKeys = MutableStateFlow<Set<String>>(emptySet())
    private val refreshTrigger = MutableStateFlow(0)
    private var viewOpenedRecorded = false

    // Re-collected on every refresh()/retry() and after a successful mutation, so the shared store re-fetches
    // rather than replaying a stale entry.
    private val settingsFeed = refreshTrigger.flatMapLatest { source.settings() }

    /** The combined surface inputs as a lifecycle-aware [StateFlow]; the projection renders it. */
    val state: StateFlow<VehicleSettingsTabState> =
        combine(settingsFeed, drafts, validation, savingKeys, resettingKeys) { settings, edits, errors, saving, resetting ->
            VehicleSettingsTabState(
                settings = settings,
                drafts = edits,
                validation = errors,
                savingKeys = saving,
                resettingKeys = resetting,
            )
        }.stateIn(
            scope = stateScope,
            started = SharingStarted.WhileSubscribed(STOP_TIMEOUT_MILLIS),
            initialValue = VehicleSettingsTabState.INITIAL,
        )

    /** Records the user's in-progress draft for [key] — the web `setDraft(e.target.value)`. */
    fun edit(
        key: String,
        value: String,
    ) {
        drafts.update { it + (key to value) }
    }

    /**
     * Validates and saves the override for [key] (web `handleSave`). A blank/invalid draft surfaces the
     * inline validation message and does not PUT; a valid draft forwards the typed value through the source,
     * then on success clears the row's draft + validation, re-fetches the feed, and surfaces the saved toast.
     */
    fun save(key: String) {
        val descriptor = descriptorForKey(key) ?: return
        validation.update { it - key }
        val draft = state.value.drafts[key] ?: effectiveDraftFor(descriptor)
        when (val parsed = parseDraft(descriptor, draft)) {
            is DraftParse.Invalid -> validation.update { it + (key to parsed.reason) }
            is DraftParse.Valid -> {
                logger.info(EVENT_SAVE, surfaceField())
                runMutation(key, savingKeys, SAVE_FEEDBACK) { source.upsert(key, parsed.value) }
            }
        }
    }

    /**
     * Reverts [key] to its inherited default (web `handleReset`). Guarded to the override case (the web
     * disables the button otherwise); on success it clears the row's draft, re-fetches the feed, and
     * surfaces the reset toast.
     */
    fun reset(key: String) {
        if (!isOverride(key)) return
        logger.info(EVENT_RESET, surfaceField())
        runMutation(key, resettingKeys, RESET_FEEDBACK) { source.reset(key) }
    }

    /** Re-fetches the settings feed after a hard error (web `refetch`); backs the retry affordance. */
    fun retry() {
        logger.info(EVENT_REFRESH, surfaceField())
        refreshTrigger.update { it + 1 }
    }

    /** Re-fetches the settings feed over the shown rows; backs the stale/offline freshness chip. */
    fun refresh() = retry()

    /**
     * Emits the one PII-safe `view.opened` diagnostic with the surface slug (P1/S11), at most once per
     * holder. Carries no setting value, VIN, or vehicle id. Call from the composable's first-composition
     * effect.
     */
    fun recordViewOpened() {
        if (viewOpenedRecorded) return
        viewOpenedRecorded = true
        VehicleSettingsTabDiagnostics.recordViewOpened(logger)
    }

    /** Runs [block], tracking [key] in [inFlight]; on success clears the draft + re-fetches, else logs + toasts. */
    private fun runMutation(
        key: String,
        inFlight: MutableStateFlow<Set<String>>,
        feedback: MutationFeedback,
        block: suspend () -> Result<Unit>,
    ) {
        inFlight.update { it + key }
        launch {
            block().fold(
                onSuccess = {
                    inFlight.update { it - key }
                    clearDraft(key)
                    refreshTrigger.update { it + 1 }
                    emitEvent(UiEvent.Message(feedback.successToast, severity = UiEvent.Severity.Success))
                },
                onFailure = { error ->
                    inFlight.update { it - key }
                    logger.warn(feedback.failEvent, surfaceField(KIND_KEY to errorKindOf(error).name))
                    emitEvent(UiEvent.Message(feedback.failToast, severity = UiEvent.Severity.Error))
                },
            )
        }
    }

    /** Clears the user's draft + validation for [key] so the row follows the refreshed effective value. */
    private fun clearDraft(key: String) {
        drafts.update { it - key }
        validation.update { it - key }
    }

    /** The effective value's draft form for [descriptor], read from the latest resolver payload. */
    private fun effectiveDraftFor(descriptor: VehicleSettingDescriptor): String =
        effectiveToDraft(descriptor, findEffectiveSetting(state.value.settings.cached, descriptor.key))

    /** Whether [key]'s effective value is a per-vehicle override (the web `source === 'override'` guard). */
    private fun isOverride(key: String): Boolean =
        findEffectiveSetting(state.value.settings.cached, key)?.source == EffectiveSettingSource.OVERRIDE

    private fun surfaceField(vararg extra: Pair<String, String>): Map<String, String> =
        mapOf(SURFACE_KEY to VehicleSettingsTabDiagnostics.SLUG, *extra)

    /** The toast keys + log event a save or reset mutation surfaces, bundled to keep the runner small. */
    private data class MutationFeedback(
        val successToast: String,
        val failToast: String,
        val failEvent: String,
    )

    companion object {
        private const val STOP_TIMEOUT_MILLIS = 5_000L
        private const val SURFACE_KEY = "surface"
        private const val KIND_KEY = "kind"
        private const val EVENT_SAVE = "vehicleSettings.save"
        private const val EVENT_SAVE_FAIL = "vehicleSettings.saveFailed"
        private const val EVENT_RESET = "vehicleSettings.reset"
        private const val EVENT_RESET_FAIL = "vehicleSettings.resetFailed"
        private const val EVENT_REFRESH = "vehicleSettings.refresh"

        private val SAVE_FEEDBACK =
            MutationFeedback(VEHICLE_SETTINGS_SAVED_KEY, VEHICLE_SETTINGS_SAVE_FAILED_KEY, EVENT_SAVE_FAIL)
        private val RESET_FEEDBACK =
            MutationFeedback(VEHICLE_SETTINGS_RESET_KEY, VEHICLE_SETTINGS_RESET_FAILED_KEY, EVENT_RESET_FAIL)

        /** A [ViewModelProvider.Factory] a host uses to construct this surface's ViewModel. */
        fun factory(
            source: VehicleSettingsTabSource,
            logger: Logger,
        ): ViewModelProvider.Factory =
            viewModelFactory {
                initializer { VehicleSettingsTabViewModel(source, logger) }
            }
    }
}
